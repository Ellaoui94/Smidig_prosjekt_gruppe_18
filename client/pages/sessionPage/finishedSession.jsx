import React, { useContext } from "react";
import { MainPageApiContext } from "../../apiContext/mainPageApiContext";
import { Link, useParams } from "react-router-dom";
import { useLoading } from "../../components/useLoading";
import { IconButton } from "@mui/material";
import { ArrowBackIosNew } from "@mui/icons-material";

// code for when you end a session, where the evaluation data of active session gets saved
export function FinishedSession() {
  const { showFinishedSession } = useContext(MainPageApiContext);
  const { sessionId } = useParams();
  const { loading, error, data } = useLoading(
    async () => await showFinishedSession({ sessionId: sessionId }),
    []
  );
  const dateFormat = {
    weekday: "long",
    month: "long",
    day: "numeric",
  };

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <>
      <IconButton
        className={"arrow-back"}
        component={Link}
        to="/main-page"
        size="large"
        aria-label="menu"
        sx={{ mr: "auto" }}
      >
        <ArrowBackIosNew />
      </IconButton>
      <br />
      <h1>{data[0].courseTitle}</h1>
      <h5 className={"session-date"}>
        {new Date(data[0].date).toLocaleDateString("no-NO", dateFormat)}
      </h5>
      <h3>{data[0].location}</h3>
      <h6 className={"session-address"}>{data[0].address}</h6>
      <hr />
      <div className={"your-evaluation-div"}>
        <h3>Din vurdering av økten</h3>
        <h5>Fokus</h5> <p>{data[0].focus}</p>
        <h5>Vurdering</h5> <p>{data[0].evaluation}</p>
        <h5>Egen kommentar</h5> <p>{data[0].comment}</p>
      </div>
      <div>
        <hr />
        <h3>Gjøremål</h3>
        {data[0].todos.length === 0 ? (
          <p>Ingen gjøremål registrert.</p>
        ) : (
          data[0].todos.map((todos) => (
            <>
              <div>
                <label>
                  <input type={"checkbox"} checked={todos.checked} readOnly />
                  {todos.todo}
                </label>
              </div>
            </>
          ))
        )}
      </div>
    </>
  );
}
