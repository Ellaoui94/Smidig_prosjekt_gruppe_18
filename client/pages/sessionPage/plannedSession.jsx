import { Link, useParams } from "react-router-dom";
import React, { useContext } from "react";
import { MainPageApiContext } from "../../mainPageApiContext";
import { useLoading } from "../../useLoading";

export function PlannedSession() {
  const { showPlannedSession } = useContext(MainPageApiContext);
  const { sessionId } = useParams();
  const { loading, error, data } = useLoading(
    async () => await showPlannedSession({ sessionId: sessionId }),
    []
  );
  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <>
      <h1>{data[0].courseTitle}</h1>
      <h5>{data[0].day + ", " + data[0].date}</h5>
      <h3>{data[0].location}</h3>
      <h6>{data[0].address}</h6>
      <div>
        <h3>Gjøremål</h3>
        {data[0].todos.map((todos) => (
          <>
            <div>
              <label>
                <input type={"checkbox"} />
                {todos}
              </label>
            </div>
          </>
        ))}
      </div>
      <Link to={"/start-session"}>Start Økt</Link>
    </>
  );
}
