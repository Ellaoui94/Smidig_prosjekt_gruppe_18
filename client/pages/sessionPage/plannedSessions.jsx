import { Link } from "react-router-dom";

export function PlannedSessions() {
  return (
    <>
      <h1>Planlagte økter</h1>
      <Link to={"/start-session"}>Start Økt</Link>
    </>
  );
}
