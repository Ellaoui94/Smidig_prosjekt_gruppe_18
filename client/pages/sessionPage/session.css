/*.borderLine {
    border: 1px solid black;
    margin: 10px
}

.itemDiv {
    border: 1px solid black;
    border-radius: 24px;
    width: 150px;
}

body {
    margin: 1em 0em 15em 0em;
    position: relative;
    overflow-x: hidden;
}

footer {
    width: 100vw;
    left: 0;
    bottom: 0;
    position: fixed;
    background-color: #D7DDD7;
}*/