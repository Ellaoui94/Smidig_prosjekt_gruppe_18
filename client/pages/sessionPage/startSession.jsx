import { Link, useNavigate } from "react-router-dom";
import { Checkbox, FormControlLabel } from "@mui/material";
import { useState } from "react";
import axios from "axios";

const evaluation = ["Bra", "Helt ok", "Dårlig"];

export function StartSession({ emailInput }) {
  console.log("before setting data.email: " + emailInput);
  const [data, setData] = useState({
    email: emailInput,
    evaluate: "",
    focus: "",
    other: "",
    finished: true,
  });

  console.log("inside startSession: " + data.email);

  const [error, setError] = useState("");
  const navigate = useNavigate();

  const handleChange = ({ currentTarget: input }) => {
    setData({ ...data, [input.name]: input.value });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const url = `${window.location.origin}/api/session/${emailInput}`;
      const { data: res } = await axios.post(url, data);
      console.log("inside handelSubmit try: " + data.email);
      navigate("/session");
      console.log(res.message);
    } catch (error) {
      if (
        error.response &&
        error.response.status >= 400 &&
        error.response.status <= 500
      ) {
        setError(error.response.data.message);
      }
    }
  };

  console.log(emailInput);
  return (
    <>
      <h1>Etter å ha trykket start på start siden kommer man hit</h1>
      <form onSubmit={handleSubmit}>
        <h2>Vurder økten</h2>
        {evaluation.map((evaluate) => (
          <div>
            {evaluate}
            <input
              type="checkbox"
              name="evaluation"
              label={"evaluation"}
              onChange={handleChange}
              value={evaluate}
            />
          </div>
        ))}

        <h2>Hvordan var fokuset under denne økten?</h2>
        {evaluation.map((evaluate) => (
          <div>
            {evaluate}
            <input
              type="checkbox"
              name="focus"
              label={"focus"}
              onChange={handleChange}
              value={evaluate}
            />
          </div>
        ))}

        <h2>Legg til kommentar</h2>
        <textarea
          name="other"
          label={"other"}
          onChange={handleChange}
          value={data.other}
        />
        <div>
          <button>Avslutt økt</button>
        </div>
      </form>
    </>
  );
}
