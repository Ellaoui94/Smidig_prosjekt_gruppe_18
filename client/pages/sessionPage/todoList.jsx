import React, { useContext, useEffect, useState } from "react";
import { MainPageApiContext } from "../../mainPageApiContext";
import { useParams } from "react-router-dom";
import { useLoading } from "../../useLoading";
import axios from "axios";

export function TodoList() {
  const { showPlannedSession, addNewTodo } = useContext(MainPageApiContext);
  const [todo, setTodo] = useState("");
  const [checked, setChecked] = useState("false");
  const [ws, setWs] = useState("");
  const [todoList, setTodoList] = useState([]);
  const [joiError, setJoiError] = useState("");
  const { sessionId } = useParams();
  const { loading, error, data } = useLoading(
    async () => await showPlannedSession({ sessionId: sessionId }),
    []
  );

  const newTodo = { todo: todo };

  useEffect(() => {
    const ws = new WebSocket(window.location.origin.replace(/^http/, "ws"));
    setWs(ws);

    // get data from websockets
    ws.onmessage = (plannedTodo) => {
      const { todo, checked } = JSON.parse(plannedTodo.data);
      setTodoList((oldState) => [...oldState, { todo, checked }]);
    };
  }, []);

  async function handleSubmit(event) {
    event.preventDefault();
    try {
      const url = `${
        window.location.origin
      }/api/session/new-todo?${sessionId}/${encodeURIComponent(
        JSON.stringify(newTodo)
      )}`;
      const { data: res } = await axios.post(url, newTodo);
      console.log(res.message);
      //send data to websockets
      ws.send(JSON.stringify({ todo, checked }));
      setTodo("");
    } catch (error) {
      if (
        error.response &&
        error.response.status >= 400 &&
        error.response.status <= 500
      ) {
        setJoiError(error.response.data.message);
      }
    }
  }

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <>
      <div>
        <h3>Gjøremål</h3>
        {joiError && <div>{joiError}</div>}
        {/* add todos from database */}
        {data[0].todos.map((todos, i) => (
          <>
            <div key={i}>
              <label>
                <input type={"checkbox"} />
                {todos.todo}
              </label>
            </div>
          </>
        ))}

        {/* add newly added todos from websockets */}
        {[...todoList].map((todos) => (
          <>
            <div>
              <label>
                <input type={"checkbox"} />
                {todos.todo}
              </label>
            </div>
          </>
        ))}

        {/* input field to add new todos */}
        <form onSubmit={handleSubmit}>
          <input
            type="text"
            id="myInput"
            onChange={(e) => setTodo(e.target.value)}
            value={todo}
            placeholder="Legg til nytt gjøremål.."
          />
          <button>Legg til</button>
        </form>
      </div>
    </>
  );
}
