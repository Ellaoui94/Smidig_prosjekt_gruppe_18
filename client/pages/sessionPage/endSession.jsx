import { Link, useNavigate, useParams } from "react-router-dom";
import { Checkbox, FormControlLabel, TextField } from "@mui/material";
import { useEffect, useState } from "react";
import axios from "axios";
import "./session.css";
import { AdapterDateFns } from "@mui/x-date-pickers/AdapterDateFns";
import { LocalizationProvider } from "@mui/x-date-pickers/LocalizationProvider";
import { DatePicker } from "@mui/x-date-pickers";

/*
   Here is the code for when you want to end an active session
 */

const evaluation = ["Bra", "Helt ok", "Dårlig"];

export function EndSession({ emailInput }) {
  const [endDateSession, setEndDateSession] = useState(null);
  const { sessionId } = useParams();

  const [data, setData] = useState({
    email: "",
    evaluation: "",
    focus: "",
    comment: "",
    stage: "finished",
    endDate: null,
  });

  console.log("inside startSession: " + emailInput);

  data.email = emailInput;
  data.endDate = endDateSession;

  const [error, setError] = useState("");
  const navigate = useNavigate();

  const handleChange = ({ currentTarget: input }) => {
    setData({ ...data, [input.name]: input.value });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      console.log("testing: " + data.evaluate);
      const url = `${window.location.origin}/api/session/update/${emailInput}`;
      const { data: res } = await axios.post(url, data);
      console.log("inside handelSubmit try: " + data.email);
      navigate("/main-page");

      //change session stage from active to finished
      const setStageUrl = `${window.location.origin}/api/session/set-session-to-finished/${sessionId}`;
      const { data: result } = await axios.post(setStageUrl, {
        stage: "finished",
      });
      console.log(result.data);
      console.log(res.message);
    } catch (error) {
      if (
        error.response &&
        error.response.status >= 400 &&
        error.response.status <= 500
      ) {
        setError(error.response.data.message);
      }
    }
  };

  return (
    <>
      <h1>Hvordan gikk studieøkten?</h1>
      <form onSubmit={handleSubmit}>
        <div className={"session-div"} style={{ backgroundColor: "white" }}>
          <h2>Vurder økten</h2>
          {evaluation.map((evaluate) => (
            <div className={"session-card-div"}>
              {evaluate}
              <input
                type="checkbox"
                name="evaluation"
                label={"evaluation"}
                onChange={handleChange}
                value={evaluate}
              />
            </div>
          ))}
        </div>

        <div className={"session-div"} style={{ backgroundColor: "white" }}>
          <h2>Hvordan var fokuset under denne økten?</h2>
          {evaluation.map((evaluate) => (
            <div className={"session-card-div"}>
              {evaluate}
              <input
                type="checkbox"
                name="focus"
                label={"focus"}
                onChange={handleChange}
                value={evaluate}
              />
            </div>
          ))}
        </div>

        <div className={"session-div"} style={{ backgroundColor: "white" }}>
          <h2>Legg til kommentar</h2>
          <textarea
            name="comment"
            label={"comment"}
            onChange={handleChange}
            value={data.comment}
          />
        </div>

        <LocalizationProvider dateAdapter={AdapterDateFns}>
          <DatePicker
            name="endDate"
            label={"endDate"}
            onChange={(newValue) => setEndDateSession(newValue)}
            value={endDateSession}
            renderInput={(params) => (
              <TextField
                style={{ background: "white" }}
                margin={"normal"}
                {...params}
              />
            )}
          />
        </LocalizationProvider>

        <div>
          <button className={"end-session-btn"}>Avslutt økt</button>
        </div>

        {error && <div>{error}</div>}
      </form>
    </>
  );
}
