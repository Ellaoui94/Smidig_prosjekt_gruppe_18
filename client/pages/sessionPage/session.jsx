import React, { useContext, useState } from "react";
import { IconButton, TextField } from "@mui/material";
import AddIcon from "@mui/icons-material/Add";
import "./session.css";
import { Link, useNavigate } from "react-router-dom";
import { UserApiContext } from "../../userApiContext";
import axios from "axios";

const subjects = ["Math", "Religion", "Physics", "History"];
const locations = ["Library", "Cafe"];
const states = ["Alone", "Invisible", "Public", "Friends only"];

export function Session({ email }) {
  const [data, setData] = useState({
    email: "",
    courseTitle: [""],
    location: "",
    studyStatus: "",
    studySessionTitle: "",
  });

  console.log("inside session: " + email);
  data.email = email;

  const [error, setError] = useState("");
  const navigate = useNavigate();

  const handleChange = ({ currentTarget: input }) => {
    setData({ ...data, [input.name]: input.value });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const url = `${window.location.origin}/api/session/${email}`;
      const { data: res } = await axios.post(url, data);
      navigate("/start-session");
      console.log(res.message);
    } catch (error) {
      if (
        error.response &&
        error.response.status >= 400 &&
        error.response.status <= 500
      ) {
        setError(error.response.data.message);
      }
    }
  };

  return (
    <div>
      <h1>Start a Session</h1>

      <form onSubmit={handleSubmit}>
        <h2>Hvilket emne vil du jobbe med?</h2>
        {subjects.map((subject) => (
          <div>
            {subject}
            <input
              type="checkbox"
              name="courseTitle"
              label={"courseTitle"}
              onChange={handleChange}
              value={subject}
            />
          </div>
        ))}

        <h2>Hvor vil du jobbe</h2>
        {locations.map((location) => (
          <div>
            {location}
            <input
              type="checkbox"
              name="location"
              label={"location"}
              onChange={handleChange}
              value={location}
            />
          </div>
        ))}

        <h2>Sett arbeidstatus for denne økten</h2>
        {states.map((status) => (
          <div>
            {status}
            <input
              type="checkbox"
              name="studyStatus"
              label={"studyStatus"}
              onChange={handleChange}
              value={status}
            />
          </div>
        ))}

        <textarea
          name="studySessionTitle"
          label={"studySessionTitle"}
          onChange={handleChange}
          value={data.studySessionTitle}
        />

        <button>Start økt</button>
      </form>
    </div>
  );
}
