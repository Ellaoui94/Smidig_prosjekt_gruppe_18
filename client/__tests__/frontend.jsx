import React from "react";
import ReactDOM from "react-dom";
import Session from "../pages/sessionPage/session";
import { FinishedSession } from "../pages/sessionPage/finishedSession";

/* This one is going to test frontend code */

describe("Session", () => {
  it("shows loading screen", () => {
    const element = document.createElement("div");
    ReactDOM.render(<FinishedSession />, element);
    expect(element.innerHTML).toMatchSnapshot();
  });
});
