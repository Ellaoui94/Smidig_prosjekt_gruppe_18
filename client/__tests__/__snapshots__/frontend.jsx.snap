// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Session shows loading screen 1`] = `"<div>Loading...</div>"`;
